File: 056/056.brute.js
// Brute force solution using BigInt primitives.
// Converting the number to string in order to find the digit sum
//  was faster than numeric method.

const digitSum = (number) => {
  /**
   * Returns the digits sum of the integer.
   * @param {int} number - to find the sum of digits for.
   * @returns {int} - the digits sum.
   */
  digits = number.toString().split('').map(Number);
  return digits.reduce((a, b) => a + b, 0);
}

const powerfulDigitSum = (limit) => {
  /**
   * Searches for the largest digit sum for numbers in the form of a ^ b.
   * @param {int} limit - the search interval, where a. b in [1, limit).
   * @returns {int} - the largest digits sum.
   */
  let largest = 0n;
  for (let a = 1n; a < limit; a++) {
    for (let b = 1n; b < limit; b++) {
      sum = digitSum(BigInt(a ** b));
      if (largest < sum) { largest = sum; }
    }
  }
  return largest;
}

// results
console.log(powerfulDigitSum(100));

File: 056/python/056.brute.py
# Brute-force solution
# It is fast -> still the approach is straightforward:
#   1. Convert the resulting integer to digit sum
#     1.1 The numeric aproach was used with mod operator.
#         Interestingly enough, the string reverse was faster.
#         And still, I decided to use numeric one.
#   2. Find the maximum digit sum.

def digits(number: int) -> int:
  """ 
  Calculates the digits sum of the given integer. \n
    Args: \n
      number: the integer to find digits sum for. \n
    Returns: \n
      The digits sum value.
  """
  while number:
    yield number % 10
    number //= 10

def powerful_digit_sum(limit):
  """
  Searches for the largest digits sum value for numbers in form of a ^ b. \n
    Args:
      limit: defines the search limit for a, b <- [1, limit).
    Returns:
      The largest digits sum value.
  """
  largest = 0
  for a in range(1, limit):
    for b in range(1, limit):
      digitsum = sum(digits(a ** b))
      largest = max(largest, digitsum)
  return largest

# results
print(powerful_digit_sum(100))

# Actually, this is the number with the largest number of digits!
# print(sum(digits(99 ** 95)))